### PR TITLE
0x_api - deduped fills - zeroex & dex_agg

### PR DESCRIPTION
**Pull request by RantumBits**
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by RantumBits** [comment]
This fails during DuneSQL check on this line: 
 `CAST(ARRAY() as array<bigint>) as trace_address`
 
I tried ` CAST(ARRAY[] as array<bigint>) as trace_address` (after testing on front-end in DuneSQL) and the test failed at run initial models. 

Could you advise on best method to address this issue?

**Comment by jeff-dude** [comment]
can you revisit this PR now that the other one is merged?

**Comment by RantumBits** [comment]
> can you revisit this PR now that the other one is merged?

updated this pr...running into an issue with the Dune SQL check. Can you advise @jeff-dude ?

**Comment by RantumBits** [comment]
all set with DuneSQL checks now @jeff-dude

File: models/zeroex/zeroex_api_fills_dex_agg.sql
Comment (jeff-dude) [inline models/zeroex/zeroex_api_fills_dex_agg.sql:1]: assuming this is the spell which will roll into `dex_aggregator.trades`, a few cleanup requests:
- can we rename the file and alias to be more consistent to others which roll into the sector spell? i.e. `alias = 'trades'` and file name `zeroex_trades`?
- can we reorg the columns to be consistent? [here](https://github.com/duneanalytics/spellbook/blob/main/models/oneinch/ethereum/oneinch_ethereum_trades.sql#L33-L56) is 1inch example

File: models/zeroex/ethereum/zeroex_ethereum_api_fills.sql [hunk 1]
```diff
@@ -455,11 +455,11 @@ SELECT
         END AS taker, -- fix the user masked by ProxyContract issue
         taker_token,
         ts.symbol AS taker_symbol,
-        maker_token,
+        maker_token, 
         ms.symbol AS maker_symbol,
         CASE WHEN lower(ts.symbol) > lower(ms.symbol) THEN concat(ms.symbol, '-', ts.symbol) ELSE concat(ts.symbol, '-', ms.symbol) END AS token_pair,
         taker_token_amount_raw / pow(10, tp.decimals) AS taker_token_amount,
-        taker_token_amount_raw,
+        cast(taker_token_amount_raw as varchar(42)) as taker_token_amount_raw,
```
Comment (jeff-dude) [inline models/zeroex/ethereum/zeroex_ethereum_api_fills.sql:462]: any particular reason this column is explicitly casted, but `maker` one isn't? also, other blockchains aren't. the column is already string in the db

File: models/zeroex/zeroex_schema.yml [hunk 2]
```diff
@@ -1,91 +1,11 @@
...
-        description: "Address which received the transaction"
-      - &trace_address
-        name: trace_address
-        description: "address of the trace within the call graph forest"
-      - &evt_index
-        name: evt_index
-        description: "the position of this event log within this transaction"
   - name: zeroex_api_fills 
     meta:
       blockchain: ethereum, optimism, polygon, arbitrum, fantom, avalanche
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:6]: add bnb?

File: models/zeroex/zeroex_schema.yml [hunk 3]
```diff
@@ -1,91 +1,11 @@
...
       project: zeroex
-      contributors: danning.sui, bakabhai993, hosuke
+      contributors: danning.sui, bakabhai993, hosuke, rantum
     config:
       tags: ['ethereum','0x','dex_aggregator','dex','aggregator']
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:10]: clean up tags? could add all blockchains or anything else relevant

File: models/zeroex/zeroex_schema.yml [hunk 4]
```diff
@@ -149,8 +91,144 @@ models:
...
+      - &evt_index
+        name: evt_index
+        description: "the position of this event log within this transaction"
+      - &blockchain
+        name: blockchain
+        description: "Blockchain which the DEX is deployed"
+
+  - name: zeroex_api_fills_deduped
+    meta: 
+      blockchain: ethereum, optimism, polygon, arbitrum, fantom, avalanche
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:109]: add bnb?

File: models/zeroex/zeroex_schema.yml [hunk 5]
```diff
@@ -149,8 +91,144 @@ models:
...
+      sector: dex
+      project: uniswap
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:111]: fix project name

File: models/zeroex/zeroex_schema.yml [hunk 6]
```diff
@@ -149,8 +91,144 @@ models:
...
+      contributors: rantum, bakabhai993
+    config:
+      tags: ['ethereum', 'optimism', 'polygon', 'uniswap','trades', 'dex', 'jeff-dude','cross-chain']
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:114]: fix tags

File: models/zeroex/zeroex_schema.yml [hunk 7]
```diff
@@ -149,8 +91,144 @@ models:
...
+      - *matcha_limit_order_flag
+      - *volume_usd
       - *tx_from
       - *tx_to
       - *evt_index
       - *blockchain
-              
\ No newline at end of file
+
+  - name: zeroex_api_fills_dex_agg
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:160]: be sure to update file name here, if changed in directory

File: models/zeroex/zeroex_schema.yml [hunk 8]
```diff
@@ -149,8 +91,144 @@ models:
...
+    meta:
+      blockchain: ethereum, optimism, polygon, arbitrum, fantom, avalanche
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:162]: add bnb?

File: models/zeroex/zeroex_schema.yml [hunk 9]
```diff
@@ -149,8 +91,144 @@ models:
...
+      project: zeroex
+      contributors: rantum, danning.sui, bakabhai993 
+    config:
+      tags: ['ethereum','0x','dex_aggregator','dex','aggregator']
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:166]: cleanup tags?

File: models/zeroex/zeroex_schema.yml [hunk 10]
```diff
@@ -149,8 +91,144 @@ models:
...
+    description: >
+        0x aggreator swaps thru 0x 0x API
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - block_date
+            - tx_hash
+            - evt_index
+      - check_seed:
+          seed_file: ref('zeroex_api_fills_dex_agg_sample')
```
Comment (jeff-dude) [inline models/zeroex/zeroex_schema.yml:176]: cleanup seed file name, if matching actual file name rename?

**Review by jeff-dude — CHANGES_REQUESTED** [review]
thanks for pushing this forward, leaving some comments below

**Review by jeff-dude — CHANGES_REQUESTED** [review]
thanks for pushing this forward, leaving some comments below

File: models/zeroex/ethereum/zeroex_ethereum_api_fills.sql (hunk 1 above)
Comment (RantumBits) [reply models/zeroex/ethereum/zeroex_ethereum_api_fills.sql]: ran intto datatype error earlier, issues seems to be resolved now

**Review by RantumBits — COMMENTED** [review]
thanks - changes made via new commits

**Review by jeff-dude — APPROVED** [review]
thank you!

assuming i didn't break anything, this should be set with the final commit i pushed to simplify the read in `zeroex.trades`. since `zeroex_api_fills_deduped` already combines all chains, we can just read from there

next step will be to follow up and integrate into `dex_aggregator.trades`. that should be as simple as adding to the list in the spell, once confident in the data

**Pull request merged**